pkg/cloudprovider/vsphere/nodemanager.go

```diff
@@ -260,81 +256,94 @@ func (nm *NodeManager) DiscoverNode(nodeID string, searchBy cm.FindVM) error {
 						},
 					)
-					found = true
-					break
+					foundInternal = true
 				}
-			} else if externalVMNetworkName != "" && strings.EqualFold(externalVMNetworkName, v.Network) {
-				for _, ip := range ips {
-					klog.V(2).Infof("Adding External IP by NetworkName: %s", ip)
+				if !foundExternal && externalNetworkSubnet != nil && externalNetworkSubnet.Contains(parsedIP) {
+					klog.V(2).Infof("Adding External IP by AddressMatching: %s", ip)
 					v1helper.AddToNodeAddresses(&addrs,
 						v1.NodeAddress{
 							Type:    v1.NodeExternalIP,
 							Address: ip,
 						},
 					)
-					found = true
-					break
+					foundExternal = true
 				}
-			} else {
-				for _, ip := range ips {
-					klog.V(2).Infof("Adding IP: %s", ip)
+
+				// then use network name
+				if !foundInternal && internalVMNetworkName != "" && strings.EqualFold(internalVMNetworkName, v.Network) {
+					klog.V(2).Infof("Adding Internal IP by NetworkName: %s", ip)
 					v1helper.AddToNodeAddresses(&addrs,
 						v1.NodeAddress{
-							Type:    v1.NodeExternalIP,
-							Address: ip,
-						}, v1.NodeAddress{
 							Type:    v1.NodeInternalIP,
 							Address: ip,
 						},
 					)
-					found = true
-					break
+					foundInternal = true
 				}
+				if !foundExternal && externalVMNetworkName != "" && strings.EqualFold(externalVMNetworkName, v.Network) {
+					klog.V(2).Infof("Adding External IP by NetworkName: %s", ip)
+					v1helper.AddToNodeAddresses(&addrs,
+						v1.NodeAddress{
+							Type:    v1.NodeExternalIP,
+							Address: ip,
+						},
+					)
+					foundExternal = true
+				}
+			}
+
+			// At least one of the Internal or External addresses has been found.
+			// Both technically need to exist for the node to function correctly.
+			// If only one was discovered, will log the warning and continue which will
+			// ultimately be visible to the end user
+			if foundInternal || foundExternal {
+				if foundInternal && !foundExternal {
+					klog.Warning("Internal address found, but external address not found. Returning what addresses were discovered.")
+				} else if !foundInternal && foundExternal {
+					klog.Warning("External address found, but internal address not found. Returning what addresses were discovered.")
+				}
+				break
 			}
 
-			if found {
+			// Neither internal or external addresses were found. This defaults to the old
+			// address selection behavior which is we only support a single address and we
+			// return the first one found
+			klog.V(5).Info("Default address selection. Single NIC, Single IP Address")
+			for _, ip := range ips {
+				klog.V(2).Infof("Adding IP: %s", ip)
+				v1helper.AddToNodeAddresses(&addrs,
+					v1.NodeAddress{
+						Type:    v1.NodeInternalIP,
+						Address: ip,
+					},
+					v1.NodeAddress{
+						Type:    v1.NodeExternalIP,
+						Address: ip,
+					},
+				)
+				foundInternal = true
+				foundExternal = true
 				break
 			}
 		}
 	}
 
-	if !found {
+	if !foundInternal || !foundExternal {
```

Comment:
I think we want this to be `!foundIInternal && !foundExternal` since it should be allowed for only 1 address to be set.

Reply:
ok